### ISSUE TITLE
Plugin system in different languages

### ISSUE THREAD

**Issue by crisidev** [koekeishiya/chunkwm#122]
Ok, this is a stretch and more a question than I real issue..

Would you think it could be possible to implement a sort of API for plugins and just speaks to chunkwm over the socket?

I am not proficient with objc, but I have a lot of experience with python and go and maybe having other languages will make writing plugins easier..

Another option I am considering is just use cgo and write native plugin in golang.

This is not a priority, just an idea I had in mind from some time ago.

**Comment by koekeishiya** [comment]
Well, it is certainly possible to establish a protocol to allow this. Many *server*, *client* types of applications do this. I'm not sure yet how we would make both systems run in parallel.

I'm sure it should be possible to create some sort of binding for the current interface though, that would let you write the majority of the code in python. 

The plugin API design did not turn out completely the way I had hoped at first, due to the rise of *macOS Sierra*. See [comment here](https://github.com/koekeishiya/chunkwm/blob/master/src/core/plugin.cpp#L279). On another note, I really dislike *objc*. I'm not a fan of *modern C++* either, which is why most of the codebase is very primitive *C++* code.

**Comment by crisidev** [comment]
Well the python binding part looks to the the simpler one from a consumer point of view. I did some cgo binding to objc and it is a real pain to encapsulate objc object inside golang one :(

**Comment by koekeishiya** [comment]
Some docs regarding integrating python: `https://docs.python.org/2/extending/embedding.html`

**Comment by koekeishiya** [comment]
@crisidev Maybe I'm completely wrong about this, but as long as you can compile shared libraries in *go*, you could theoretically do everything from *go* without a binding, as long as the shared library exports the necessary format that *chunkwm* expects upon calling *dlopen*, *dlsym*.

**Comment by crisidev** [comment]
@koekeishiya correct. The only part needed is an header file with the function which need to be exported to chunkwm (aka what you defined in the API pdf). These functions will have to follow the API for parameter and returns, using cgo. The rest can be pure go. The problem is if you use complex types as arguments or return values. These need to be translate into go types using cgo and vice versa to c++ types. This is what I understood about cgo writing PAM modules.

**Comment by koekeishiya** [comment]
@crisidev The types passed between *chunkwm* and plugins are as follows:

passed to plugins during init / event:
[chunkwm_api](https://github.com/koekeishiya/chunkwm/blob/master/src/api/plugin_cvar.h#L24)
[chunkwm_payload](https://github.com/koekeishiya/chunkwm/blob/master/src/api/plugin_export.h#L4)
[macos_window](https://github.com/koekeishiya/chunkwm/blob/master/src/common/accessibility/window.h#L29)
[macos_application](https://github.com/koekeishiya/chunkwm/blob/master/src/common/accessibility/application.h#L9)

other from */common/*:
[macos_space](https://github.com/koekeishiya/chunkwm/blob/master/src/common/accessibility/display.h#L36)
[macos_display](https://github.com/koekeishiya/chunkwm/blob/master/src/common/accessibility/display.h#L60)

If code from */common/* is used, *std::vector* is also utilized for some parts.

**Comment by crisidev** [comment]
I'll try to take a look to it next week and maybe try to build a query plugin on golang if you are not already working on it in objc.

**Comment by crisidev** [comment]
I took a look over the types used and
* chunkwm_api should be easy
* chunkwm_payload should be easy

All the macos_* are more complex since the types contain other internal complex types which I think are referred to OSX library type like AXWindow and so on. Anyway, I'll try to start working on a wrapper ASAP.

**Comment by koekeishiya** [comment]
@crisidev I have not started on anything related to queries yet.

Info about the macOS types that can come in handy. Some of the types requires some more investigation I suppose. 

From *ApplicationServices/Frameworks/HIServices/AXUIElement.h*:
**AXObserverRef**:
```
An AXObserverRef is a CFTypeRef.
typedef struct CF_BRIDGED_TYPE(id) __AXObserver *AXObserverRef;
```
**AXUIElementRef**: 
```
An AXObserverRef is a CFTypeRef.
typedef const struct CF_BRIDGED_TYPE(id) __AXUIElement *AXUIElementRef;
```

**ProcessSerialNumber**:
```
struct ProcessSerialNumber {
   unsigned long highLongOfPSN;
   unsigned long lowLongOfPSN;
};
```

**CGFloat**:
```
When you build for a 64-bit CPU, the CGFloat type is a 64-bit, 
IEEE double-precision floating point type, equivalent to the Double type. 

When you build for a 32-bit CPU, the CGFloat type is a 32-bit,
IEEE single-precision floating point type, equivalent to the Float type.
```

**CGPoint**:
```
struct CGPoint {
    CGFloat x;
    CGFloat y;
};
typedef struct CGPoint CGPoint;
```

**CGSize**:
```
struct CGSize {
    CGFloat width;
    CGFloat height;
};
typedef struct CGSize CGSize;
```

**Comment by crisidev** [comment]
So far I was able to implement a golang mock of PLUGIN_MAIN_FUNC and PLUGIN_VOID_FUNC.

The rest is more complex. I am expanding all the macros with clang++ -E src/core/chunkwm.mm to see if it is doable :)

**Comment by crisidev** [comment]
I did some research and looks like cgo does not support building .mm files. 

Between this and some of the macros (namely CHUNKWM_PLUGIN) which do not expand properly in cgo, I see building a golang interface VERY VERY complex :(

I will do other researches anyway. Regarding the query plugin, it will be way more easy to build it in objc. If we want to pursue the golang / python plugin way, IMHO API interaction via socket is the better approach. Otherwise I think some rethink will be needed inside chunkwm core and plugin internal structure. That said I am no expert in objc and I did not used cgo that much to be 100% sure I am not saying BS :D

**Comment by koekeishiya** [comment]
Hm the macros just define a couple of pretty primitive functions, so you don't need to actually use the macros.

I don't know how *cgo* or *go* works in general, so I can't really provide much help here.

You don't really need to touch any *objc* though. The plugin interface can be used from more or less a simple *C89* application. If you understand and can read / write basic *C* programs, that should be enough. 

Depending on what the plugin is going to do, you might need support for interfacing with isolated *objc* code, the way I do to create borders or access *NSWorkspace*. None of the *objc* specific types are exposed outside their *.mm* file.

**Comment by crisidev** [comment]
The problem are the types passes to every primitive function used to register a new plugin. It is not matter of objc / c / c++ but types.

Cgo can be used in two ways:
1) import and use current C / C+ code. This is very complex because of what I said above. Cgo is not able to compile mm code and you are using std::vector in headers, which is totally unsupported. Simply it does not build without changing chunkwm quite a lot
2) rewrite the plugin API in go and export functions to C, following the plugin API. This mean reimplement all the types passed as arguments and returned by exported functions in go. Basically it is rewriting the full plugin part in go.

**Comment by koekeishiya** [comment]
Due to the fact that we have to interact with windows / applications through their AXUIElementRef types, any plugin that wants to actually interact with the system needs to be able to represent this type.

The easiest solution from my perspective, would thus be to leave the chunkwm C-API intact.
You then build a "control" protocol on top of the command system that chunkwm exposes to their plugins.

Roughly what this might look like

```
command -> socket -> chunkwm
chunkwm -> plugin->run -> "custom dispatch protocol"
"custom dispatch protcol" -> socket -> <entry to code in whatever language>

<code in whatever language> -> socket -> chunkwm -> plugin->run -> "custom command protocol"
```

Lets take the example of moving and resizing windows.
The plugin that chunkwm loads subscribes to the window-related events per normal.
You now create wrapper functions that can move a window by identiying it uniquely with its window id, (pid if application and so on). 

You dispatch the same event, replacing the window data structure with the format your protocol requires, (probably use window id as identifier), and pass the new primitive data to the plugin written in *go* or something else.

When you want to move a window from the *go* plugin, dispatch a message following your protocol back to the *C-plugin*, through the chunkwm command system. The *C-plugin* reads understands your protocol and maps the window id (or whatever info) back to a macos_window struct and can call the AX-API functions on it.

**Comment by koekeishiya** [comment]
@crisidev 

Regarding the above approach. Instead of defining a text protocol through the command system, we could have a minimal C-plugin that interfaces with chunkwm. This plugin could then *dlopen* some other shared library (a *go* library), which defines its own ABI.

Events received from chunkwm to the minimal C-plugin could then call handlers inside the shared library with the appropriate arguments.

The reason I don't want to do this directly in chunkwm, is that I do not want plugins to have to interact with the AXUIElementRefs through chunkwm.

**Comment by crisidev** [comment]
@koekeishiya I think this is a good approach!

**Comment by koekeishiya** [comment]
If you are interested in trying something like the above, that would be cool. I don't have any experience with *go* etc, so I can only contribute with the C-parts that you may need help with.

I think the dlopen approach is better than the text-based protocol.

**Comment by crisidev** [comment]
I would love to do it. This week I am really busy with thou.. I can start poking around next week. We should first define an API for the C part I think.

**Comment by lego** [comment]
This is a pretty interesting idea :). I've begun taking a stab at making an RPC plugin with a minimal Python plugin.

I haven't worked out the specifics of what the best way create an interface in the other language (given all of the C function calls to fetch or manipulate stuff), but I've been thinking about the border plugin as a milestone. 

(If you're curious, you can check out the WIP code [here](https://github.com/koekeishiya/chunkwm/compare/master...lego:joey/rpc-plugin))

**Comment by splintah** [comment]
Hi,
I've begun making Rust 'bindings' to the chunkwm plugin structure. You can find the repo [here](https://github.com/splintah/chunkwm-rs), and the template repo [here](https://github.com/splintah/chunkwm-rs-template).

@koekeishiya, it would be nice if the methods defined in common could be `extern "C"`. That would make it easier to generate bindings to those (for now I've made a copy, but I would like to add chunkwm as a submodule). Shall I make a PR?

The design currently is: compile the Rust lib to a C-compatible static lib, then 'import' this in `plugin.cpp`, and compile that to a shared library. I will try to make it Rust only, but for now this works.

**Comment by koekeishiya** [comment]
> @koekeishiya, it would be nice if the methods defined in common could be extern "C". That would make it easier to generate bindings to those (for now I've made a copy, but I would like to add chunkwm as a submodule). Shall I make a PR?

I'll take a look at this after I'm finished with https://github.com/koekeishiya/chunkwm/issues/315

**Comment by koekeishiya** [comment]
@splintah 

I've look a bit through the */common/* code, and notice two issues with making them all extern. First, a couple (1 or 2) functions make use of **std::vector**. This is very trivial to change, and is not really a problem. The other issue is more work and involves quite heavy usage of function overload (especially the space related functions).

Could you perhaps be more specific regarding which parts are interesting / important.

**Comment by splintah** [comment]
Hmm... I just realised: I could also create an `extern "C"` wrapper for the functions I want to use.
It's not that much work; and you needn't change your API.

Otherwise: I think there isn't a good way of creating bindings for overloaded functions, except for creating a wrapper.

Conclusion: I'm going to create a wrapper, so probably no change needed on your/this end.

**Comment by koekeishiya** [comment]
Aight sounds good. Let me know if you have any questions regarding the code. I haven't kept the `plugin api docs` up to date, and comments in code are very much lacking, with the exception of the `chunkwm core`, which is not really something you need to care about to get this working.

**Comment by koekeishiya** [comment]
@splintah 

You may want to submodule a specific *chunkwm tag* in case I do breaking changes on master, so that you do not accidentally break bindings for the latest stable version due to a  pre-emptive submodule update. 

I suppose this depends on how you handle versioning on your side.

**Comment by splintah** [comment]
Thank you, I've just done it :)

**Comment by splintah** [comment]
I've been trying to remove all C code in the plugin template (not that I hate C), which has worked out quite great.
But there was one thing I stumbled upon, which is the disability to export `static`s in Rust.
So, @koekeishiya, might it be possible to change the exported `plugin_details Exports` to a function like `plugin_details GetExports()`?

**Comment by koekeishiya** [comment]
@splintah 

You should be able to do the `plugin_details Exports` in  rust like this:

```
#[repr(C)]
pub struct plugin_details {
    number: i32,
}

#[no_mangle]
pub static Exports: plugin_details = plugin_details {
    number: 10,
};
```

and just fill in the struct members and the initialization values instead of the dummy ones that I've placed above.

Then compile the rust code to a shared library: `rustc --crate-type cdylib`

The following c-code can then load the rust library and retrieve the static Exports variable that was declared in global scope ( in rust ):
```
#include <stdio.h>
#include <stdlib.h>
#include <dlfcn.h>

struct MyStruct {
    int32_t number;
};

int main()
{
    void *handle = dlopen("libmain.dylib", RTLD_LAZY);
    if (handle) {
        struct MyStruct *my_struct = (struct MyStruct *) dlsym(handle, "Exports");
        printf("%d\n", my_struct->number);
        dlclose(handle);
    }
}
```

**Comment by splintah** [comment]
Owh, hmm, I thought exporting `static`s in Rust was unsupported.
Well, thank you!

**Comment by splintah** [comment]
The `#[no_mangle] pub static Export` approach works, but the problem is that `*const c_char`s (`const char *` in C) can't be created as a `static` from regular Rust `&'static str`s. The current implementation uses `fn name() -> &'static str` (and the same for `version`) to specify the plugin's name, and since [`const fn`](https://github.com/rust-lang/rust/issues/24111) isn't stable yet, those cannot be used for statically exporting the strings.
When `const fn` will be stable, `std::ffi::CString::new(...)` can be used to create the C-compatible pointer.

One way to fix this problem is to use a `struct` contains a `*const u8`, like explained [here](https://github.com/rust-lang/rust/issues/24111). The `create_bridge!` macro could receive these strings as arguments.
But the problem is that strings should be declared like `b"string contents\0"`. The macro usage would look like
```rust
create_bridge!(Plugin, b"plugin name\0", b"version\0")
```
which isn't great...

Another option I thought of was using [`lazy_static!`](https://docs.rs/lazy_static/1.0.0/lazy_static/), but this does not work since the static isn't actually a static, but a singleton that gets initialised at runtime.

**Comment by koekeishiya** [comment]
Hmm I would've thought something like this would work

```
const fname: *const u8 = &b"filename\0"[0];
const pname: *const u8 = &b"pluginname\0"[0];
const pvers: *const u8 = &b"0.0.1\0"[0];

#[repr(C)]
pub struct plugin_details {
    api_version: i32,
    filename: *const u8,
    plugin_name: *const u8,
    plugin_version: *const u8,
}

#[no_mangle]
pub static Exports: plugin_details = plugin_details {
    api_version: 10,
    filename: fname,
    plugin_name: pname,
    plugin_version: pvers
};
```

I'm not very familiar with rust though, so yeah

**Comment by splintah** [comment]
Well, using a pointer _does_ work, like:
```rust
const fname: *const u8 = b"file name\0" as *const u8;
```
but a nul character must be added, and it must be defined as bytes.

**Comment by koekeishiya** [comment]
I'll agree that is looks a bit messy like that, but if it works I personally think that is the best approach.

The alternative is that chunkwm does a *dlsym* to get a function that returns a singleton structure of type *plugin_details*, which would require us to do a *dlsym* followed by a function call, rathern than just the *dlsym* that is done today. I'd prefer not having to do that.

**Comment by splintah** [comment]
> The alternative is that chunkwm does a dlsym to get a function that returns a singleton structure of type plugin_details, which would require us to do a dlsym followed by a function call, rathern than just the dlsym that is done today. I'd prefer not having to do that.

Yeah, I understand. I will try to come up with other alternatives, and otherwise I will probably use the `b"\0"` method.
